node/src/main/scala/coop/rchain/node/configuration/toml/TomlConfiguration.scala

```diff
@@ -0,0 +1,91 @@
+package coop.rchain.node.configuration.toml
+
+import java.io.File
+import java.nio.file.{Path, Paths}
+
+import scala.io.Source
+import scala.util.Try
+
+import cats.syntax.either._
+
+import coop.rchain.comm.PeerNode
+import coop.rchain.shared.Resources._
+
+import toml._
+import toml.Codecs._
+
+object TomlConfiguration {
+  private implicit val bootstrapAddressCodec: Codec[PeerNode] =
+    Codec {
+      case (Value.Str(uri), _) =>
+        PeerNode
+          .parse(uri)
+          .map(u => Right(u))
+          .getOrElse(Left((Nil, "can't parse the rnode bootstrap address")))
+      case _ => Left((Nil, "the rnode bootstrap address should be a string"))
+    }
+  private implicit val pathCodec: Codec[Path] =
+    Codec {
+      case (Value.Str(uri), _) =>
+        Try(Paths.get(uri)).toEither.leftMap(_ => (Nil, s"Can't parse the path $uri"))
+      case _ => Left((Nil, "A path must be a string"))
+    }
+  private implicit val boolCodec: Codec[Boolean] = Codec {
+    case (Value.Bool(value), _) => Right(value)
+    case (value, _) =>
+      Left((List.empty, s"Bool expected, $value provided"))
+  }
+
+  def from(toml: String): Either[String, Configuration] =
+    Toml.parse(toml) match {
+      case Left(error) => Either.left(s"Failed to parse TOML string: $error")
+      case Right(ast)  => from(ast)
+    }
+
+  def from(ast: Value.Tbl): Either[String, Configuration] =
+    Toml.parseAs[Configuration](rewriteKeysToCamelCase(ast)) match {
+      case Left((_, error)) => Either.left(s"Failed to parse TOML AST: $error")
+      case Right(root)      => Either.right(root)
+    }
+
+  def from(file: File): Either[String, Configuration] =
```

Comment:
Yes, `from(file)` calls `from(string)` calls `from(ast)`. It would be possible to squash the code into one method. But for the sake of readability there are three methods.